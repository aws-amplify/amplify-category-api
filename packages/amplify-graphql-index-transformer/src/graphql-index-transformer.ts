import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InvalidDirectiveError,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { IndexDirective } from '@aws-amplify/graphql-directives';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isListType, isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { generateKeyAndQueryNameForConfig, validateNotSelfReferencing } from './utils';

/**
 *
 */
export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];

  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', IndexDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as IndexDirectiveConfiguration,
      generateGetArgumentsInput(context.transformParameters),
    );

    /**
     * Impute Optional Fields
     * Start with sort key fields since both the name and queryField will work sort key fields into the generated name if necessary.
     */
    args.sortKeyFields = getOrGenerateDefaultSortKeyFields(args);
    args.name = getOrGenerateDefaultName(args);
    args.queryField = getOrGenerateDefaultQueryField(context, args);
    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

/**
 * Return the name if provided in our args, else
 * compute the name based on the field name, and sortKeyFields.
 */
const getOrGenerateDefaultName = (config: IndexDirectiveConfiguration): string => {
  const indexNameRegex = /^[A-Za-z0-9_\-\.]{3,255}$/;
  if (config.name) {
    if (!indexNameRegex.test(config.name)) {
      throw new Error(
        `The indexName is invalid. It should be between 3 and 255 characters. Only A–Z, a–z, 0–9, underscore characters, hyphens, and periods allowed.`,
      );
    }
    return config.name;
  }

  if (config.name === null) {
    throw new Error('Explicit null value not allowed for name field on @index');
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * Return the queryField if provided in our args, else
 * compute the queryField based on the field name, and sortKeyFields if the feature flag is enabled.
 */
const getOrGenerateDefaultQueryField = (
  context: TransformerSchemaVisitStepContextProvider,
  config: IndexDirectiveConfiguration,
): string | null => {
  const autoIndexQueryNamesIsEnabled = context.transformParameters.enableAutoIndexQueryNames;
  // Any explicit null will take effect, if enableAutoIndexQueryNames and no queryField is provide set to null for consistency
  if (config.queryField === null || (!autoIndexQueryNamesIsEnabled && !config.queryField)) {
    return null;
  }

  if (config.queryField) {
    return config.queryField;
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * sortKeyFields are optional so default to empty list,
 * if we get a raw object just wrap in an array,
 * else return the list which was provided correctly.
 */
const getOrGenerateDefaultSortKeyFields = (config: IndexDirectiveConfiguration): string[] => {
  if (!config.sortKeyFields) {
    return [];
  }
  if (!Array.isArray(config.sortKeyFields)) {
    return [config.sortKeyFields];
  }
  return config.sortKeyFields;
};

const validate = (config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void => {
  const { name, object, field, sortKeyFields } = config;
  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find((directive) => directive.name.value === 'model');

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${IndexDirective.name} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        const hasSortFields = peerDirective.arguments!.some(
          (arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0,
        );
        config.primaryKeyField = objectField;

        if (!hasSortFields && objectField.name.value === field.name.value) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === IndexDirective.name &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${IndexDirective.name} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }

    for (const peerDirective of objectField.directives!) {
      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0);

      if (
        !ctx.transformParameters.secondaryKeyAsGSI &&
        !hasSortFields &&
        objectField == config.primaryKeyField &&
        objectField.name.value === field.name.value
      ) {
        throw new InvalidDirectiveError(
          `Invalid @index '${name}'. You may not create an index where the partition key ` +
            'is the same as that of the primary key unless the index has a sort field. ' +
            'You cannot have a local secondary index without a sort key in the index.',
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums) || isListType(field.type)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums) || isListType(sortField.type)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
};
