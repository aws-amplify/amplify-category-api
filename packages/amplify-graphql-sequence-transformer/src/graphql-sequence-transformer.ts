import { SequenceDirectiveConfiguration } from './types';
import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InputObjectDefinitionWrapper,
  InvalidDirectiveError,
  isPostgresModel,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { SequenceDirective } from '@aws-amplify/graphql-directives';
import { DirectiveNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { getBaseType, ModelResourceIDs } from 'graphql-transformer-common';

export const ERR_NOT_MODEL = 'The @sequence directive may only be added to object definitions annotated with @model.';
export const ERR_NOT_INT = 'The @sequence directive may only be applied to integer fields';
export const ERR_NOT_POSTGRES = 'The @sequence directive may only be applied to Postgres datasources';
export const ERR_ARGC = 'The @sequence directive does not take any arguments';

const validateModelDirective = (config: SequenceDirectiveConfiguration): void => {
  const modelDirective = config.object.directives!.find((dir) => dir.name.value === 'model');
  if (!modelDirective) {
    throw new InvalidDirectiveError(ERR_NOT_MODEL);
  }
};

const validateFieldType = (config: SequenceDirectiveConfiguration): void => {
  const baseTypeName = getBaseType(config.field.type);
  if (baseTypeName !== 'Int') {
    throw new InvalidDirectiveError(ERR_NOT_INT);
  }
};

const validateDatasourceType = (ctx: TransformerSchemaVisitStepContextProvider, config: SequenceDirectiveConfiguration): void => {
  const isPostgres = isPostgresModel(ctx, config.object.name.value);
  if (!isPostgres) {
    throw new InvalidDirectiveError(ERR_NOT_POSTGRES);
  }
};

const validate = (ctx: TransformerSchemaVisitStepContextProvider, config: SequenceDirectiveConfiguration): void => {
  validateModelDirective(config);
  validateFieldType(config);
  validateDatasourceType(ctx, config);
};

export class SequenceTransformer extends TransformerPluginBase {
  private directiveMap = new Map<string, SequenceDirectiveConfiguration[]>();

  constructor() {
    super('amplify-sequence-transformer', SequenceDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    ctx: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const config = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as SequenceDirectiveConfiguration,
      generateGetArgumentsInput(ctx.transformParameters),
    );
    validate(ctx, config);

    if (!this.directiveMap.has(parent.name.value)) {
      this.directiveMap.set(parent.name.value, []);
    }

    this.directiveMap.get(parent.name.value)!.push(config);
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    for (const typeName of this.directiveMap.keys()) {
      const name = ModelResourceIDs.ModelCreateInputObjectName(typeName);
      for (const config of this.directiveMap.get(typeName)!) {
        const inputObject = InputObjectDefinitionWrapper.fromObject(name, config.object, ctx.inputDocument);
        const appliedField = inputObject.fields.find((f) => f.name === config.field.name.value);
        appliedField!.makeNullable();

        ctx.output.updateInput(inputObject.serialize());
      }
    }
  };
}
