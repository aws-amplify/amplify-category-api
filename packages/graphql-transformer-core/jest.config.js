const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
