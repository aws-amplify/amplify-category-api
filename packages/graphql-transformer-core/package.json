{
  "name": "graphql-transformer-core",
  "version": "8.2.4",
  "description": "A framework to transform from GraphQL SDL to AWS cloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "rimraf ./lib ./tsconfig.tsbuildinfo && tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^10.3.0",
    "graphql": "^15.5.0",
    "graphql-transformer-common": "4.27.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@aws-amplify/amplify-cli-core": "^4.2.10",
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^18.17.0",
    "mock-fs": "5.2.0",
    "@types/mock-fs": "4.13.4"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 67,
        "functions": 33,
        "lines": 48
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
