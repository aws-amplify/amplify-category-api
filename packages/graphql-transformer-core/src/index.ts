import './polyfills/Object.assign';
import { print } from 'graphql';
import { DeploymentResources } from './DeploymentResources';
import { TransformerContext, MappingParameters } from './TransformerContext';
import { Transformer } from './Transformer';
import { ITransformer } from './ITransformer';
import { GraphQLTransform } from './GraphQLTransform';
import { collectDirectiveNames, collectDirectivesByType, collectDirectivesByTypeNames } from './collectDirectives';
import { stripDirectives } from './stripDirectives';
import {
  buildProject as buildAPIProject,
  uploadDeployment as uploadAPIProject,
  migrateAPIProject,
  revertAPIMigration,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
} from './util/amplifyUtils';
import {
  readSchema as readProjectSchema,
  loadProject as readProjectConfiguration,
  loadConfig as readTransformerConfiguration,
  writeConfig as writeTransformerConfiguration,
} from './util/transformConfig';
import { EXTRA_DIRECTIVES_DOCUMENT } from './validation';

export * from './errors';
export * from './util';
export { getTableNameForModel } from './tableNameMap';

/**
 * Returns the set of directives that are supported by AppSync service
 */
export function getAppSyncServiceExtraDirectives(): string {
  return print(EXTRA_DIRECTIVES_DOCUMENT);
}
export { FeatureFlagProvider } from './FeatureFlags';

export {
  GraphQLTransform,
  TransformerContext,
  MappingParameters,
  Transformer,
  ITransformer,
  collectDirectiveNames,
  collectDirectivesByType,
  collectDirectivesByTypeNames,
  stripDirectives,
  buildAPIProject,
  migrateAPIProject,
  uploadAPIProject,
  readProjectSchema,
  readProjectConfiguration,
  readTransformerConfiguration,
  writeTransformerConfiguration,
  revertAPIMigration,
  DeploymentResources,
  CLOUDFORMATION_FILE_NAME,
  PARAMETERS_FILE_NAME,
};
// No-op change to trigger publish
