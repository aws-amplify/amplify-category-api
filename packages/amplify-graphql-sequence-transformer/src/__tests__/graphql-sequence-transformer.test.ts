import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { mockSqlDataSourceStrategy, testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { parse } from 'graphql';
import { SequenceTransformer, ERR_NOT_INT, ERR_NOT_MODEL, ERR_NOT_POSTGRES, ERR_ARGC } from '../graphql-sequence-transformer';
import { constructDataSourceStrategies, POSTGRES_DB_TYPE, validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';

describe('SequenceTransformer:', () => {
  it('throws if @sequence is used in a non-@model type', () => {
    const schema = `
      type Test {
        id: ID! @sequence
        name: String
      }`;

    expect(() =>
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new SequenceTransformer()],
      }),
    ).toThrow(ERR_NOT_MODEL);
  });

  it.each([
    { strategy: mockSqlDataSourceStrategy() },
    // TODO: DTE Mock DynamoDB?
  ])('throws if @sequence is used on a non Postgres datasource', ({ strategy }) => {
    const schema = `
      type CoffeeQueue @model {
        id: ID! @primaryKey
        orderNumber: Int! @sequence
        name: String
      }`;
    expect(() => {
      testTransform({
        schema: schema,
        transformers: [new ModelTransformer(), new SequenceTransformer(), new PrimaryKeyTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
      });
    }).toThrow(ERR_NOT_POSTGRES);
  });

  it.each([
    { typeStr: 'Boolean' },
    { typeStr: 'AWSJSON' },
    { typeStr: 'AWSDate' },
    { typeStr: 'AWSDateTime' },
    { typeStr: 'AWSTime' },
    { typeStr: 'AWSTime' },
    { typeStr: 'AWSURL' },
    { typeStr: 'AWSPhone' },
    { typeStr: 'AWSIPAddress' },
  ])('throws if @sequence is used on non-int types', ({ typeStr }) => {
    expect(() => {
      const schema = `
      type Test @model {
        id: ID!
        value: ${typeStr} @sequence
      }
    `;
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new SequenceTransformer()],
      });
    }).toThrow(ERR_NOT_INT);
  });

  it('should successfully transform simple valid schema', async () => {
    const postgresStrategy = mockSqlDataSourceStrategy({ dbType: POSTGRES_DB_TYPE });

    const inputSchema = `
      type CoffeeQueue @model {
        id: ID! @primaryKey
        orderNumber: Int! @sequence
        name: String
      }
    `;
    const out = testTransform({
      schema: inputSchema,
      transformers: [new ModelTransformer(), new SequenceTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(inputSchema, postgresStrategy),
    });
    expect(out).toBeDefined();
    expect(out.schema).toMatchSnapshot();

    const schema = parse(out.schema);
    validateModelSchema(schema);
  });
});
