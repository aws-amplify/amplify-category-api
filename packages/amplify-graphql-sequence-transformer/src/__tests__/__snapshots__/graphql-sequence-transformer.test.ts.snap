// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceTransformer: should successfully transform simple valid schema 1`] = `
"type CoffeeQueue {
  id: ID!
  orderNumber: Int!
  name: String
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelCoffeeQueueConnection {
  items: [CoffeeQueue]!
  nextToken: String
}

input ModelCoffeeQueueFilterInput {
  id: ModelIDInput
  orderNumber: ModelIntInput
  name: ModelStringInput
  and: [ModelCoffeeQueueFilterInput]
  or: [ModelCoffeeQueueFilterInput]
  not: ModelCoffeeQueueFilterInput
}

type Query {
  getCoffeeQueue(id: ID!): CoffeeQueue
  listCoffeeQueues(id: ID, filter: ModelCoffeeQueueFilterInput, limit: Int, nextToken: String, sortDirection: ModelSortDirection): ModelCoffeeQueueConnection
}

input ModelCoffeeQueueConditionInput {
  orderNumber: ModelIntInput
  name: ModelStringInput
  and: [ModelCoffeeQueueConditionInput]
  or: [ModelCoffeeQueueConditionInput]
  not: ModelCoffeeQueueConditionInput
}

input CreateCoffeeQueueInput {
  id: ID
  orderNumber: Int
  name: String
}

input UpdateCoffeeQueueInput {
  id: ID!
  orderNumber: Int
  name: String
}

input DeleteCoffeeQueueInput {
  id: ID!
}

type Mutation {
  createCoffeeQueue(input: CreateCoffeeQueueInput!, condition: ModelCoffeeQueueConditionInput): CoffeeQueue
  updateCoffeeQueue(input: UpdateCoffeeQueueInput!, condition: ModelCoffeeQueueConditionInput): CoffeeQueue
  deleteCoffeeQueue(input: DeleteCoffeeQueueInput!, condition: ModelCoffeeQueueConditionInput): CoffeeQueue
}

input ModelSubscriptionCoffeeQueueFilterInput {
  id: ModelSubscriptionIDInput
  orderNumber: ModelSubscriptionIntInput
  name: ModelSubscriptionStringInput
  and: [ModelSubscriptionCoffeeQueueFilterInput]
  or: [ModelSubscriptionCoffeeQueueFilterInput]
}

type Subscription {
  onCreateCoffeeQueue(filter: ModelSubscriptionCoffeeQueueFilterInput): CoffeeQueue @aws_subscribe(mutations: [\\"createCoffeeQueue\\"])
  onUpdateCoffeeQueue(filter: ModelSubscriptionCoffeeQueueFilterInput): CoffeeQueue @aws_subscribe(mutations: [\\"updateCoffeeQueue\\"])
  onDeleteCoffeeQueue(filter: ModelSubscriptionCoffeeQueueFilterInput): CoffeeQueue @aws_subscribe(mutations: [\\"deleteCoffeeQueue\\"])
}
"
`;
